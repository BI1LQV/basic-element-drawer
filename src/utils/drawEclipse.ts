const { round } = Math
export function eclipseCore(Rx: number, Ry: number) {
  let Rx2 = Rx * Rx
  let Ry2 = Ry * Ry
  let twoRx2 = 2 * Rx2
  let twoRy2 = 2 * Ry2
  let p
  let x = 0
  let y = Ry
  let px = 0
  let py = twoRx2 * y
  let res = []
  // Plot the initial point in each quadrant
  res.push([x, y])
  /* Region 1 */
  p = round(Ry2 - (Rx2 * Ry) + (0.25 * Rx2))
  while (px < py) {
    x++
    px += twoRy2
    if (p < 0) {
      p += Ry2 + px
    } else {
      y--
      py -= twoRx2
      p += Ry2 + px - py
    }
    res.push([x, y])
  }
  /* Region 2 */
  p = round(Ry2 * (x + 0.5) * (x + 0.5) + Rx2 * (y - 1) * (y - 1) - Rx2 * Ry2)
  while (y > 0) {
    y--
    py -= twoRx2
    if (p > 0) {
      p += Rx2 - py
    } else {
      x++
      px += twoRx2
      p += Rx2 - py + px
    }
    res.push([x, y])
  }
  return res
}

export function* centerEclipse(x: number, y: number, Rx: number, Ry: number) {
  let core = eclipseCore(Rx, Ry)
  yield core.map(([x0, y0]) => [x + y0, y + x0])
  yield core.map(([x0, y0]) => [x - y0, y + x0])
  yield core.map(([x0, y0]) => [x + y0, y - x0])
  yield core.map(([x0, y0]) => [x - y0, y - x0])
}
