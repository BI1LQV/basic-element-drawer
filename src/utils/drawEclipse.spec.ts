import { describe, expect, it } from "vitest"
import { centerEclipse, eclipseCore } from "./drawEclipse"

describe("draw circ", () => {
  it("should works", () => {
    expect(eclipseCore(3, 2)).toMatchSnapshot()
    expect(Array.from(centerEclipse(5, 5, 2, 3))).toMatchSnapshot()
  })
})
