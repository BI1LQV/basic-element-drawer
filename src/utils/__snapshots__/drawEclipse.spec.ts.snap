// Vitest Snapshot v1

exports[`draw circ > should works 1`] = `
[
  [
    0,
    2,
  ],
  [
    1,
    2,
  ],
  [
    2,
    1,
  ],
  [
    3,
    0,
  ],
]
`;

exports[`draw circ > should works 2`] = `
[
  [
    [
      8,
      5,
    ],
    [
      8,
      6,
    ],
    [
      7,
      7,
    ],
    [
      6,
      7,
    ],
    [
      5,
      7,
    ],
  ],
  [
    [
      2,
      5,
    ],
    [
      2,
      6,
    ],
    [
      3,
      7,
    ],
    [
      4,
      7,
    ],
    [
      5,
      7,
    ],
  ],
  [
    [
      8,
      5,
    ],
    [
      8,
      4,
    ],
    [
      7,
      3,
    ],
    [
      6,
      3,
    ],
    [
      5,
      3,
    ],
  ],
  [
    [
      2,
      5,
    ],
    [
      2,
      4,
    ],
    [
      3,
      3,
    ],
    [
      4,
      3,
    ],
    [
      5,
      3,
    ],
  ],
]
`;
